File: bindings/dotnet/Ztok/Format.cs
// Format.cs — vocab-format enum + auto-detect helper.

using System;
using System.IO;
using System.Text;

namespace Ztok;

/// <summary>
/// Auto-detected on-disk vocab format. Mirrors <c>ztok_format</c> in
/// <c>include/ztok.h</c>; values are stable and additive.
/// </summary>
public enum Format : uint
{
    /// <summary>Unrecognized or unreadable file.</summary>
    Unknown = 0,
    /// <summary>OpenAI-style <c>.tiktoken</c> byte-level BPE vocab.</summary>
    Tiktoken = 1,
    /// <summary>HuggingFace <c>tokenizer.json</c> (BPE or WordPiece).</summary>
    HfJson = 2,
    /// <summary>SentencePiece <c>.model</c> (Unigram).</summary>
    SentencePiece = 3,
    /// <summary>ztok native <c>.ztm</c> TokenMonster format.</summary>
    Ztm = 4,
    /// <summary>Mistral <c>tekken.json</c> (tiktoken-style with extra config).</summary>
    Tekken = 5,
}

/// <summary>Normalizer kind (mirrors <c>ztok_normalizer_kind</c>).</summary>
public enum Normalizer : uint
{
    /// <summary>No normalization.</summary>
    Identity = 0,
    /// <summary>Unicode NFC normalization.</summary>
    Nfc = 1,
    /// <summary>Unicode NFD normalization.</summary>
    Nfd = 2,
    /// <summary>Unicode NFKC normalization.</summary>
    Nfkc = 3,
    /// <summary>Unicode NFKD normalization.</summary>
    Nfkd = 4,
    /// <summary>Byte-level normalization (GPT-2 / cl100k style).</summary>
    ByteLevel = 5,
}

/// <summary>Pre-tokenizer kind (mirrors <c>ztok_pretok_kind</c>).</summary>
public enum PreTokenizer : uint
{
    /// <summary>No pre-tokenization (treat input as a single span).</summary>
    Identity = 0,
    /// <summary>OpenAI cl100k_base regex split.</summary>
    Cl100k = 1,
}

/// <summary>Decoder kind (mirrors <c>ztok_decoder_kind</c>).</summary>
public enum Decoder : uint
{
    /// <summary>Concatenate token bytes verbatim.</summary>
    Concat = 0,
    /// <summary>WordPiece-style decoding (handles "##" continuation markers).</summary>
    WordPiece = 1,
    /// <summary>Byte-level decoding (inverse of byte-level normalization).</summary>
    ByteLevel = 2,
}

/// <summary>
/// Overlay channel kind (mirrors <c>ztok_overlay_kind</c>). Pass a set of
/// these to <see cref="Pipeline.EncodeWithOverlays"/> to request per-token
/// annotation channels aligned 1:1 with the id stream. Cheap channels
/// (<see cref="ByteStart"/> / <see cref="ByteEnd"/> / <see cref="Boundary"/>
/// / <see cref="Provenance"/>) carry encoder-derived values; domain
/// channels (<see cref="Opcode"/> / <see cref="Operand"/> /
/// <see cref="SymbolRef"/> / <see cref="Hunk"/>) come back zero-filled
/// until a domain plugin populates them.
/// </summary>
public enum OverlayKind : uint
{
    /// <summary>Original-input byte offset where the token's span starts.</summary>
    ByteStart = 0,
    /// <summary>Original-input byte offset where the token's span ends (exclusive).</summary>
    ByteEnd = 1,
    /// <summary>Boundary bitset: 0x1 chunk-start, 0x2 codepoint-start.</summary>
    Boundary = 2,
    /// <summary>Domain: normalized opcode class (zero-filled without a plugin).</summary>
    Opcode = 3,
    /// <summary>Domain: normalized operand class (zero-filled without a plugin).</summary>
    Operand = 4,
    /// <summary>Domain: symbol-table index, 0 = none (zero-filled without a plugin).</summary>
    SymbolRef = 5,
    /// <summary>Domain: diff-hunk id (zero-filled without a plugin).</summary>
    Hunk = 6,
    /// <summary>Provenance: 0 = model text, 1 = special token.</summary>
    Provenance = 7,
}

/// <summary>Format-detection helpers and libztok version probe.</summary>
public static class ZtokLibrary
{
    /// <summary>
    /// libztok version string (e.g. <c>"1.23.0"</c>).
    /// </summary>
    public static string Version
    {
        get
        {
            var raw = Native.Version();
            return Native.PtrToStringUtf8(raw) ?? string.Empty;
        }
    }

    /// <summary>
    /// Sniff <paramref name="path"/> for a known tokenizer format.
    /// Best-effort: any I/O error or unrecognized magic returns
    /// <see cref="Format.Unknown"/>.
    /// </summary>
    public static Format DetectFormat(string path)
    {
        ArgumentNullException.ThrowIfNull(path);
        return (Format)Native.AutoDetect(path);
    }
}

File: bindings/swift/Sources/Ztok/Format.swift
// Format — Swift mirror of `ztok_format`.
//
// Returned by `Pipeline.detectFormat(_:)` (best-effort sniffer). The C
// ABI guarantees that any I/O error or unrecognized magic maps to
// `.unknown` rather than surfacing as an error, so the Swift API does
// the same.

import CZtok
import Foundation

/// Auto-detected on-disk vocab format. Values mirror `ztok_format`
/// integers exactly so future additions stay non-breaking — unknown
/// codes fall through to `.unknown`.
public enum Format: UInt32, Sendable, CaseIterable {
    /// Unrecognized or unreadable file.
    case unknown = 0
    /// OpenAI-style `.tiktoken` byte-level BPE vocab.
    case tiktoken = 1
    /// HuggingFace `tokenizer.json` (BPE or WordPiece).
    case hfJson = 2
    /// SentencePiece `.model` (Unigram).
    case sentencePiece = 3
    /// ztok native `.ztm` TokenMonster format.
    case ztm = 4
    /// Mistral `tekken.json` (tiktoken-style with extra config).
    case tekken = 5
    /// RWKV "World" vocab (`rwkv_vocab_v20230424.txt`) — greedy
    /// longest-match byte trie.
    case rwkv = 6

    /// Build a `Format` from the raw `ztok_format` integer. Unknown
    /// codes map to `.unknown` — matches the C ABI's "best-effort,
    /// never errors" contract.
    internal static func fromCode(_ code: UInt32) -> Format {
        return Format(rawValue: code) ?? .unknown
    }
}

/// Unicode / byte-level normalizer kind (mirrors `ztok_normalizer_kind`).
public enum Normalizer: UInt32, Sendable {
    case identity = 0
    case nfc = 1
    case nfd = 2
    case nfkc = 3
    case nfkd = 4
    case byteLevel = 5
}

/// Pre-tokenizer kind (mirrors `ztok_pretok_kind`).
public enum PreTokenizer: UInt32, Sendable {
    case identity = 0
    case cl100k = 1
    /// Mistral Tekken pattern (used by `Pipeline.fromTekken`).
    case tekken = 2
}

/// Decoder kind (mirrors `ztok_decoder_kind`).
public enum Decoder: UInt32, Sendable {
    case concat = 0
    case wordPiece = 1
    case byteLevel = 2
}

/// Overlay channel kind (mirrors `ztok_overlay_kind`).
///
/// Pass a list of these to `Pipeline.encodeWithOverlays(_:channels:)` to
/// request per-token annotation channels aligned 1:1 with the id stream.
/// Cheap channels (`.byteStart` / `.byteEnd` / `.boundary` /
/// `.provenance`) carry encoder-derived values; domain channels
/// (`.opcode` / `.operand` / `.symbolRef` / `.hunk`) come back
/// zero-filled until a domain plugin populates them.
public enum OverlayKind: UInt32, Sendable, Hashable, CaseIterable {
    /// Original-input byte offset where the token's span starts.
    case byteStart = 0
    /// Original-input byte offset where the token's span ends (exclusive).
    case byteEnd = 1
    /// Boundary bitset: 0x1 chunk-start, 0x2 codepoint-start.
    case boundary = 2
    /// Domain: normalized opcode class (zero-filled without a plugin).
    case opcode = 3
    /// Domain: normalized operand class (zero-filled without a plugin).
    case operand = 4
    /// Domain: symbol-table index, 0 = none (zero-filled without a plugin).
    case symbolRef = 5
    /// Domain: diff-hunk id (zero-filled without a plugin).
    case hunk = 6
    /// Provenance: 0 = model text, 1 = special token.
    case provenance = 7
}

/// Overlay domain (mirrors `ztok_overlay_domain`).
///
/// Selects which domain normalizer populates the domain overlay channels
/// (`.opcode` / `.operand` / `.symbolRef` / `.hunk`). Pass to
/// `Pipeline.setOverlayDomain(_:)`. `.none` (the default) leaves those
/// channels zero-filled; `.x86_64` decodes the input as x86-64 machine code.
public enum OverlayDomain: UInt32, Sendable {
    /// No domain normalizer; domain channels stay zero-filled (default).
    case none = 0
    /// Decode the input as x86-64 machine code.
    case x86_64 = 1
}

/// Boundary mode for `Pipeline.chunk(_:maxTokens:overlap:boundary:)`
/// (mirrors `ztok_chunk_boundary`). Selects where chunk edges are
/// allowed to fall.
public enum ChunkBoundary: UInt32, Sendable {
    /// Pure token-count windows (default).
    case token = 0
    /// Snap to a UTF-8 codepoint boundary.
    case codepoint = 1
    /// Snap to a whitespace word boundary.
    case word = 2
    /// Snap to a dictionary word boundary (CJK/Thai/...).
    case wordDict = 3
    /// Snap to a sentence boundary.
    case sentence = 4
    /// Snap to a paragraph break (`\n\n`).
    case paragraph = 5
}

/// Optional configuration for pipeline constructors.
///
/// The model kind is implicit (set by the constructor used: BPE,
/// WordPiece, Unigram, Monster, or byte_id). Pre-tokenizer defaults
/// differ between loaders — `Pipeline.fromTiktoken` overrides to
/// `.cl100k` unless the caller passes a custom config.
public struct PipelineConfig: Sendable {
    public var normalizer: Normalizer
    public var preTokenizer: PreTokenizer
    public var decoder: Decoder

    public init(
        normalizer: Normalizer = .identity,
        preTokenizer: PreTokenizer = .identity,
        decoder: Decoder = .concat
    ) {
        self.normalizer = normalizer
        self.preTokenizer = preTokenizer
        self.decoder = decoder
    }

    internal func toC() -> ztok_pipeline_config {
        // ztok_model_kind only exposes BYTE_ID from ztok_pipeline_new;
        // the file constructors ignore the model field anyway because
        // they hard-wire BPE / WordPiece / Unigram / Monster.
        return ztok_pipeline_config(
            normalizer: ztok_normalizer_kind(rawValue: normalizer.rawValue),
            pre_tokenizer: ztok_pretok_kind(rawValue: preTokenizer.rawValue),
            model: ZTOK_MODEL_BYTE_ID,
            decoder: ztok_decoder_kind(rawValue: decoder.rawValue)
        )
    }
}
